pkg/installer/filebuilder.go

```diff
@@ -0,0 +1,237 @@
+	if !config.Spec.IncludeEtcd {
+		fs, err = createDirAndFiles(fs, fsData)
+		if err != nil {
+			return fs, err
+		}
+		return fs, nil
+	}
+
+	etcdSubDirs := make(map[string]map[string][]byte)
+
+	// build etcd/operator
+	etcdOpFiles, err := newFileBuilder(config.Spec.Install.EtcdOperatorYaml, "", pluginversion.EtcdOperatorLatestSupportedImageURL(), etcdOperatorFile, config.Spec.GetOperatorNamespace()).createFileWithKustPair(clientConfig)
+	if err != nil {
+		return fs, err
+	}
+	etcdSubDirs[operatorDir] = etcdOpFiles
+
+	// build etcd/cluster
+	etcdClusterFiles, err := newFileBuilder(config.Spec.Install.EtcdClusterYaml, pluginversion.EtcdClusterLatestSupportedURL(), pluginversion.EtcdOperatorLatestSupportedImageURL(), etcdClusterFile, config.Spec.GetOperatorNamespace()).createFileWithKustPair(clientConfig)
+	if err != nil {
+		return fs, err
+	}
+	etcdSubDirs[clusterDir] = etcdClusterFiles
+
+	fsData[etcdDir] = etcdSubDirs
+	fs, err = createDirAndFiles(fs, fsData)
+	if err != nil {
+		return fs, err
+	}
+
+	return fs, nil
+}
+
+// createFileWithKustPair creates a map of two files (file name to file data).
+//
+// The first file is that which has its address stored in fileBuilder as a
+// local path, github release URL or manifests image repo
+//
+// The second file is the kustomization.yaml created from scratch.
+// It's contents, to begin with are simply:
+//
+// resources:
+// - <filename>
+//
+func (fb *fileBuilder) createFileWithKustPair(config *rest.Config) (map[string][]byte, error) {
+	files, err := fb.createFileWithData(config)
+	if err != nil {
+		return files, err
+	}
+
+	kustYamlContents, err := pluginutils.SetFieldInManifest(kustTemp, fmt.Sprintf("%s%s%s", "[", fb.fileName, "]"), "resources", "")
+	if err != nil {
+		return files, err
+	}
+
+	files[kustomizationFile] = []byte(kustYamlContents)
+
+	return files, nil
+}
+
+// createFileWithData returns a map with a single entry of [filename][filecontent]
+func (fb *fileBuilder) createFileWithData(config *rest.Config) (map[string][]byte, error) {
+	file := make(map[string][]byte)
+	yamlContents, err := fb.readOrPullManifest(config)
+	if err != nil {
+		return file, err
+	}
+	file[fb.fileName] = []byte(yamlContents)
+
+	return file, nil
+}
+
+// readOrPullManifest returns a string of the manifest from path, url or image provided
+func (fb *fileBuilder) readOrPullManifest(config *rest.Config) (string, error) {
+	location := fb.yamlPath
+	if location == "" {
+		location = fb.yamlImage
+	}
+	if location == "" {
+		location = fb.yamlUrl
+	}
+	if location == "" {
+		return "", errors.WithStack(errors.New("manifest location not set"))
+	}
+
+	if isDockerRepo(location) {
+		contents, err := fetchImageAndExtractFileFromTarball(location, fb.fileName)
+		if err == nil {
+			return contents, nil
+		}
+		// could not get file from image (may not exist on provided version)
+		// default to release URL
+		location = fb.yamlUrl
+	}
+
+	if util.IsURL(location) {
```

Comment:
Maybe an `else if` helps understand the relation between those two.

Reply:
Just realised an `else if` will mean we bypass the `location = fb.yamlUrl`.

There may be a case where the docker repo exists, but the manifest we want does not (eg `operator-manifests:v2.5.0` will not have `storageos-cluster.yaml`) so then we need to default to the github release url instead and pull from there.

Reply:
ohh i see.